packages/@aws-cdk/aws-elasticsearch/lib/domain.ts

```diff
@@ -0,0 +1,1503 @@
+      defaultInstanceType;
+    const dedicatedMasterCount = props.capacity?.masterNodes ?? 0;
+    const dedicatedMasterEnabled = dedicatedMasterCount > 0;
+
+    const instanceType =
+      props.capacity?.dataNodeInstanceType?.toLowerCase() ??
+      defaultInstanceType;
+    const instanceCount = props.capacity?.dataNodes ?? 1;
+
+    const availabilityZoneCount =
+      props.zoneAwareness?.availabilityZoneCount ?? 2;
+
+    if (![2, 3].includes(availabilityZoneCount)) {
+      throw new Error('Invalid zone awareness configuration; availabilityZoneCount must be 2 or 3');
+    }
+
+    const zoneAwarenessEnabled =
+      props.zoneAwareness?.enabled ??
+      props.zoneAwareness?.availabilityZoneCount != null;
+
+    if ([dedicatedMasterType, instanceType].some(t => !t.endsWith('.elasticsearch'))) {
+      throw new Error('Master and data node instance types must end with ".elasticsearch".');
+    }
+
+    const elasticsearchVersion = props.version.version;
+    const elasticsearchVersionNum = parseVersion(elasticsearchVersion);
+
+    function parseVersion(version: string): number {
+      const firstDot = version.indexOf('.');
+
+      if (firstDot < 1) {
+        throw new Error(`Invalid Elasticsearch version: ${version}. Version string needs to start with major and minor version (x.y).`);
+      }
+
+      const secondDot = version.indexOf('.', firstDot + 1);
+
+      try {
+        if (secondDot == -1) {
+          return parseFloat(version);
+        } else {
+          return parseFloat(version.substring(0, secondDot));
+        }
+      } catch (error) {
+        throw new Error(`Invalid Elasticsearch version: ${version}. Version string needs to start with major and minor version (x.y).`);
+      }
+    }
+
+    if (
+      elasticsearchVersionNum <= 7.7 &&
+      ![
+        1.5, 2.3, 5.1, 5.3, 5.5, 5.6, 6.0,
+        6.2, 6.3, 6.4, 6.5, 6.7, 6.8, 7.1, 7.4,
+        7.7,
+      ].includes(elasticsearchVersionNum)
+    ) {
+      throw new Error(`Unknown Elasticsearch version: ${elasticsearchVersion}`);
+    }
+
+    const masterUserArn = props.fineGrainedAccessControl?.masterUserArn;
+    const masterUserName = props.fineGrainedAccessControl?.masterUserName;
+
+    const advancedSecurityEnabled = (masterUserArn ?? masterUserName) != null;
+    const internalUserDatabaseEnabled = masterUserName != null;
+    const masterUserPasswordString = props.fineGrainedAccessControl?.masterUserPasswordSecret?.toString();
+    const createMasterUserPasswordSecret = (): string => {
+      return new secretsmanager.Secret(this, id, {
+        generateSecretString: {
+          secretStringTemplate: JSON.stringify({
+            username: masterUserName,
+          }),
+          generateStringKey: 'password',
+        },
+      })
+        .secretValueFromJson('password')
+        .toString();
+    };
+    const masterUserPassword =
+      masterUserPasswordString ??
+      internalUserDatabaseEnabled ? createMasterUserPasswordSecret() : undefined;
```

Comment:
I did like your version better so I changed it :smile:

I have also added a field to domain to expose the password fiel.d

Reply:
👍  - Lets also add a test for this?

Reply:
I mean the fact that the user configured password is considered

Reply:
I think the existing test actually showed that the user's password wasn't used. I have added a test for dynamically generated password and changed the existing test so that the provided password is used.